File: src/main/java/dev/isxander/evergreenhud/elements/impl/ElementPotionHUD.java
/*
 * Copyright (C) isXander [2019 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/isXander/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package dev.isxander.evergreenhud.elements.impl;

import dev.isxander.evergreenhud.elements.ElementData;
import dev.isxander.evergreenhud.elements.RenderOrigin;
import dev.isxander.evergreenhud.elements.type.BackgroundElement;
import dev.isxander.evergreenhud.elements.type.TextElement;
import dev.isxander.evergreenhud.settings.impl.BooleanSetting;
import dev.isxander.evergreenhud.settings.impl.EnumSetting;
import dev.isxander.evergreenhud.settings.impl.IntegerSetting;
import dev.isxander.evergreenhud.settings.impl.StringSetting;
import dev.isxander.evergreenhud.utils.RomanNumeral;
import dev.isxander.xanderlib.utils.GuiUtils;
import dev.isxander.xanderlib.utils.Resolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.InventoryEffectRenderer;
import net.minecraft.client.resources.I18n;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;
import net.minecraft.util.EnumChatFormatting;
import net.minecraft.util.ResourceLocation;

import java.awt.*;
import java.util.*;
import java.util.List;
import java.util.stream.Collectors;

public class ElementPotionHUD extends BackgroundElement {

    public static final int ICON_SIZE = 18;
    public static final ResourceLocation EFFECTS_RESOURCE = new ResourceLocation("textures/gui/container/inventory.png");

    public IntegerSetting titleTextR;
    public IntegerSetting titleTextG;
    public IntegerSetting titleTextB;
    public EnumSetting<TextElement.TextMode> titleTextMode;
    public BooleanSetting titleTextChroma;
    public BooleanSetting titleTextBold;
    public BooleanSetting titleTextUnderline;
    public BooleanSetting titleTextItalic;
    public BooleanSetting showTitle;
    public BooleanSetting showAmplifier;
    public EnumSetting<AmplifierMode> amplifierText;
    public BooleanSetting showAmplifierLevelOne;

    public IntegerSetting timeTextR;
    public IntegerSetting timeTextG;
    public IntegerSetting timeTextB;
    public EnumSetting<TextElement.TextMode> timeTextMode;
    public BooleanSetting timeTextChroma;
    public BooleanSetting timeTextUnderline;
    public BooleanSetting timeTextItalic;
    public BooleanSetting showTime;
    public IntegerSetting blinkingTime;
    public IntegerSetting blinkingSpeed;

    public BooleanSetting showSpeed;
    public BooleanSetting showSlowness;
    public BooleanSetting showHaste;
    public BooleanSetting showMiningFatigue;
    public BooleanSetting showStrength;
    public BooleanSetting showJumpBoost;
    public BooleanSetting showNausea;
    public BooleanSetting showRegeneration;
    public BooleanSetting showResistance;
    public BooleanSetting showFireResistance;
    public BooleanSetting showWaterBreathing;
    public BooleanSetting showInvisibility;
    public BooleanSetting showBlindness;
    public BooleanSetting showNightVision;
    public BooleanSetting showHunger;
    public BooleanSetting showWeakness;
    public BooleanSetting showPoison;
    public BooleanSetting showWither;
    public BooleanSetting showHealthBoost;
    public BooleanSetting showAbsorption;
    public BooleanSetting showSaturation;

    public BooleanSetting showPermanent;
    public StringSetting permanentText;

    public BooleanSetting showIcon;
    public EnumSetting<VerticalAlignmentMode> verticalAlign;
    public EnumSetting<SortingMode> sortingMode;
    public IntegerSetting verticalSpacing;
    public BooleanSetting overwriteIER;


    private float width = 10f;
    private float height = 10f;

    @Override
    public void initialise() {
        addSettings(titleTextR            = new IntegerSetting("Title Red",        "Title", "The red channel of the title text.",   255, 0, 255, ""));
        addSettings(titleTextG            = new IntegerSetting("Title Green",      "Title", "The green channel of the title text.", 255, 0, 255, ""));
        addSettings(titleTextB            = new IntegerSetting("Title Blue",       "Title", "The blue channel of the title text.",  255,  0, 255, ""));
        addSettings(titleTextMode         = new EnumSetting<>( "Title Mode",       "Title", "How should the text be rendered.", TextElement.TextMode.SHADOW));
        addSettings(titleTextChroma       = new BooleanSetting("Title Chroma",     "Title", "If the color of the title is a multicolored mess.",        false));
        addSettings(titleTextBold         = new BooleanSetting("Title Bold",       "Title", "If the title is bold.",                                    true));
        addSettings(titleTextUnderline    = new BooleanSetting("Title Underlined", "Title", "If the title is underlined.",                              false));
        addSettings(showTitle             = new BooleanSetting("Show Title",       "Title", "Show the title.",                                          true));
        addSettings(titleTextItalic       = new BooleanSetting("Title Italic",     "Title", "If the title is italic.",                                  false));
        addSettings(showAmplifier         = new BooleanSetting("Show Amplifier",   "Title", "Show the potion amplifier.",                               true));
        addSettings(showAmplifierLevelOne = new BooleanSetting("Show LVL 1",       "Title", "Whether or not the element displays amplifier level one.", false));
        addSettings(amplifierText         = new EnumSetting<>( "Amplifier Text",   "Title", "How the amplifier should be displayed.", AmplifierMode.ARABIC));

        addSettings(timeTextR             = new IntegerSetting("Time Red",               "Time", "The red channel of the time text.",   255, 0, 255, ""));
        addSettings(timeTextG             = new IntegerSetting("Time Green",             "Time", "The green channel of the time text.", 255, 0, 255, ""));
        addSettings(timeTextB             = new IntegerSetting("Time Blue",              "Time", "The blue channel of the time text.",  180, 0, 255, ""));
        addSettings(timeTextMode          = new EnumSetting<>( "Time Mode",              "Time", "How should the text be rendered.", TextElement.TextMode.SHADOW));
        addSettings(timeTextChroma        = new BooleanSetting("Time Chroma",            "Time", "If the color of the time is a multicolored mess.", false));
        addSettings(timeTextUnderline     = new BooleanSetting("Time Underlined",        "Time", "If the time is underlined.",                       false));
        addSettings(timeTextItalic        = new BooleanSetting("Time Italic",            "Time", "If the time is italic.",                           false));
        addSettings(showTime              = new BooleanSetting("Show Time",              "Time", "Show the time.",                                   true));
        addSettings(showPermanent         = new BooleanSetting("Show Permanent Effects", "Time", "Show permanent effects on the HUD.",               true));
        addSettings(permanentText         = new StringSetting( "Permanent Text",         "Time", "What text is displayed in place of the time if the potion is permanent.", "**:**"));
        addSettings(blinkingTime          = new IntegerSetting("Blinking Time",          "Time", "When the duration should start to blink.", 5,  0,  30, " secs"));
        addSettings(blinkingSpeed         = new IntegerSetting("Blinking Speed",         "Time", "How fast the time blinks.",                30, 10, 45, ""));

        addSettings(showSpeed             = new BooleanSetting("Speed",          "Whitelist", "Display the speed effect on the HUD.",           true));
        addSettings(showSlowness          = new BooleanSetting("Slowness",       "Whitelist", "Display the slowness effect on the HUD.",        true));
        addSettings(showHaste             = new BooleanSetting("Haste",          "Whitelist", "Display the haste effect on the HUD.",           true));
        addSettings(showMiningFatigue     = new BooleanSetting("Mining Fatigue", "Whitelist", "Display the mining fatigue effect on the HUD.",  true));
        addSettings(showStrength          = new BooleanSetting("Strength",       "Whitelist", "Display the strength effect on the HUD.",        true));
        addSettings(showJumpBoost         = new BooleanSetting("Jump Boost",     "Whitelist", "Display the jump boost effect on the HUD.",      true));
        addSettings(showNausea            = new BooleanSetting("Nausea",         "Whitelist", "Display the nausea effect on the HUD.",          true));
        addSettings(showRegeneration      = new BooleanSetting("Regeneration",   "Whitelist", "Display the regeneration effect on the HUD.",    true));
        addSettings(showResistance        = new BooleanSetting("Resistance",     "Whitelist", "Display the resistance effect on the HUD.",      true));
        addSettings(showFireResistance    = new BooleanSetting("Fire Res",       "Whitelist", "Display the fire resistance effect on the HUD.", true));
        addSettings(showWaterBreathing    = new BooleanSetting("Water Breathe",  "Whitelist", "Display the water breathing effect on the HUD.", true));
        addSettings(showInvisibility      = new BooleanSetting("Invisibility",   "Whitelist", "Display the invisibility effect on the HUD.",    true));
        addSettings(showBlindness         = new BooleanSetting("Blindness",      "Whitelist", "Display the blindness effect on the HUD.",       true));
        addSettings(showNightVision       = new BooleanSetting("Night Vision",   "Whitelist", "Display the night vision effect on the HUD.",    true));
        addSettings(showHunger            = new BooleanSetting("Hunger",         "Whitelist", "Display the hunger effect on the HUD.",          true));
        addSettings(showWeakness          = new BooleanSetting("Weakness",       "Whitelist", "Display the weakness effect on the HUD.",        true));
        addSettings(showPoison            = new BooleanSetting("Poison",         "Whitelist", "Display the poison effect on the HUD.",          true));
        addSettings(showWither            = new BooleanSetting("Wither",         "Whitelist", "Display the wither effect on the HUD.",          true));
        addSettings(showHealthBoost       = new BooleanSetting("Health Boost",   "Whitelist", "Display the health boost effect on the HUD.",    true));
        addSettings(showAbsorption        = new BooleanSetting("Absorption",     "Whitelist", "Display the absorption effect on the HUD.",      true));
        addSettings(showSaturation        = new BooleanSetting("Saturation",     "Whitelist", "Display the saturation effect on the HUD.",      true));

        addSettings(showIcon              = new BooleanSetting("Show Icon",           "Other", "Show the status effect icon.", true));
        addSettings(verticalAlign         = new EnumSetting<>( "Vertical Align",      "Other", "How the effects should be aligned vertically.",   VerticalAlignmentMode.DOWN));
        addSettings(sortingMode           = new EnumSetting<>( "Sort",                "Other", "In what way the effects shall be sorted.",        SortingMode.DURATION));
        addSettings(verticalSpacing       = new IntegerSetting("Vertical Spacing",    "Other", "How far apart each potion will be.", 2, 1, 9, " px"));
        addSettings(overwriteIER          = new BooleanSetting("Overwrite Inventory", "Other", "Overwrites the vanilla PotionHUD in the inventory to instead display this element.", false));
    }

    @Override
    public void onAdded() {
        super.onAdded();

        PotionHUDTracker.INSTANCE.instances.add(this);
    }

    @Override
    public void onRemoved() {
        super.onRemoved();

        PotionHUDTracker.INSTANCE.instances.remove(this);
    }

    @Override
    protected ElementData metadata() {
        return new ElementData("PotionHUD", "Displays all current potion effects.", "Advanced");
    }

    @Override
    public void render(float partialTicks, int origin) {
        // we would be rendering it twice
        if (origin == RenderOrigin.HUD && overwriteIER.get() && mc.currentScreen instanceof InventoryEffectRenderer)
            return;

        GlStateManager.color(1f, 1f, 1f, 1f);
        GlStateManager.disableLighting();

        float x = getPosition().getRawX(Resolution.get());
        float y = getPosition().getRawY(Resolution.get());

        List<PotionEffect> potionEffects = new ArrayList<>();
        if (mc.thePlayer != null) potionEffects.addAll(mc.thePlayer.getActivePotionEffects());
        potionEffects = filterEffects(potionEffects);

        if (potionEffects.isEmpty()) {
            if (origin == RenderOrigin.GUI) {
                potionEffects.add(new PotionEffect(Potion.damageBoost.id, 50, 4));
                potionEffects.add(new PotionEffect(Potion.moveSpeed.id, 5000, 2));
                PotionEffect effect = new PotionEffect(Potion.saturation.id, 10000000, 0);
                effect.setPotionDurationMax(true);
                potionEffects.add(effect);
                potionEffects.add(new PotionEffect(Potion.absorption.id, 242, 1));
                potionEffects = filterEffects(potionEffects);
            } else {
                return;
            }
        }

        if (sortingMode.get() == SortingMode.ALPHABETICAL) {
            potionEffects.sort(Comparator.comparing(o -> I18n.format(o.getEffectName())));
        } else if (sortingMode.get() == SortingMode.DURATION) {
            potionEffects.sort(Comparator.comparingInt(PotionEffect::getDuration));
            Collections.reverse(potionEffects);
        }
        if (verticalAlign.get() == VerticalAlignmentMode.UP) Collections.reverse(potionEffects);

        float yOff = 0;
        float xOff = 0;
        final int yAmt = ICON_SIZE + verticalSpacing.get();

        this.height = (potionEffects.size() * yAmt) - verticalSpacing.get();
        super.render(partialTicks, origin);
        this.width = 10f;

        GlStateManager.pushMatrix();
        GlStateManager.scale(getPosition().getScale(), getPosition().getScale(), 1);
        for (PotionEffect effect : potionEffects) {
            Potion potion = Potion.potionTypes[effect.getPotionID()];
            if (!potion.shouldRender(effect)) continue;

            float iconX = x;
            iconX /= getPosition().getScale();

            GlStateManager.color(1f, 1f, 1f, 1f);

            if (showIcon.get()) {
                mc.getTextureManager().bindTexture(EFFECTS_RESOURCE);
                drawTexturedModalRectF(iconX, (y + yOff) / getPosition().getScale(), potion.getStatusIconIndex() % 8 * 18, 198 + potion.getStatusIconIndex() / 8 * 18, 18, 18);
                xOff = ICON_SIZE * getPosition().getScale();
                this.width = Math.max(this.width, xOff / getPosition().getScale());
            }

            if (showTitle.get()) {
                if (showIcon.get())
                    xOff = (ICON_SIZE + 4) * getPosition().getScale();

                StringBuilder titleSb = new StringBuilder();
                if (titleTextBold.get()) titleSb.append(EnumChatFormatting.BOLD);
                if (titleTextItalic.get()) titleSb.append(EnumChatFormatting.ITALIC);
                if (titleTextUnderline.get()) titleSb.append(EnumChatFormatting.UNDERLINE);
                titleSb.append(I18n.format(potion.getName()));
                int amplifier = Math.max(1, effect.getAmplifier() + 1);
                if (showAmplifier.get() && (amplifier != 1 || showAmplifierLevelOne.get())) {
                    titleSb.append(" ");
                    if (amplifierText.get() == AmplifierMode.ROMAN) titleSb.append(RomanNumeral.INSTANCE.getCache(amplifier));
                    else titleSb.append(amplifier);
                }
                String builtTitle = titleSb.toString();

                int titleWidth = mc.fontRendererObj.getStringWidth(builtTitle);
                width = Math.max(width, (xOff / getPosition().getScale()) + titleWidth);

                float titleX = x + xOff;
                titleX /= getPosition().getScale();

                float titleY = y + yOff;
                if (!showTime.get())
                    titleY += mc.fontRendererObj.FONT_HEIGHT / 2f;
                titleY /= getPosition().getScale();


                GuiUtils.drawString(mc.fontRendererObj, builtTitle,
                        titleX, titleY, titleTextMode.get() == TextElement.TextMode.SHADOW, titleTextMode.get() == TextElement.TextMode.BORDER, titleTextChroma.get(), false, new Color(titleTextR.get(), titleTextG.get(), titleTextB.get()).getRGB());

            }

            if (showTime.get()) {
                if (showIcon.get())
                    xOff = (ICON_SIZE + 4) * getPosition().getScale();

                StringBuilder timeSb = new StringBuilder();
                if (timeTextItalic.get()) timeSb.append(EnumChatFormatting.ITALIC);
                if (timeTextUnderline.get()) timeSb.append(EnumChatFormatting.UNDERLINE);
                if (effect.getIsPotionDurationMax()) timeSb.append(permanentText.get());
                else timeSb.append(Potion.getDurationString(effect));
                String builtTime = timeSb.toString();

                int timeWidth = mc.fontRendererObj.getStringWidth(builtTime);
                width = Math.max(width, (xOff / getPosition().getScale()) + timeWidth);

                float timeX = x + xOff;
                timeX /= getPosition().getScale();

                float timeY = y + yOff + (mc.fontRendererObj.FONT_HEIGHT * getPosition().getScale()) + 1;
                if (!showTitle.get())
                    timeY -= mc.fontRendererObj.FONT_HEIGHT / 2f;
                timeY /= getPosition().getScale();

                if (effect.getDuration() / 20f > blinkingTime.get() || effect.getDuration() % (50 - blinkingSpeed.get()) <= (50 - blinkingSpeed.get()) / 2f) {
                    GuiUtils.drawString(mc.fontRendererObj, builtTime,
                            timeX, timeY, timeTextMode.get() == TextElement.TextMode.SHADOW, timeTextMode.get() == TextElement.TextMode.BORDER, timeTextChroma.get(), false, new Color(timeTextR.get(), timeTextG.get(), timeTextB.get()).getRGB());

                }
            }

            yOff += yAmt * getPosition().getScale();
        }
        GlStateManager.popMatrix();
    }

    private List<PotionEffect> filterEffects(List<PotionEffect> potionEffects) {
        return potionEffects.stream().filter(effect -> !((!showSpeed.get() && effect.getPotionID() == Potion.moveSpeed.id)
                || (!showSlowness.get() && effect.getPotionID() == Potion.moveSlowdown.id)
                || (!showHaste.get() && effect.getPotionID() == Potion.digSpeed.id)
                || (!showMiningFatigue.get() && effect.getPotionID() == Potion.digSlowdown.id)
                || (!showStrength.get() && effect.getPotionID() == Potion.damageBoost.id)
                || (!showJumpBoost.get() && effect.getPotionID() == Potion.jump.id)
                || (!showNausea.get() && effect.getPotionID() == Potion.confusion.id)
                || (!showRegeneration.get() && effect.getPotionID() == Potion.regeneration.id)
                || (!showResistance.get() && effect.getPotionID() == Potion.resistance.id)
                || (!showFireResistance.get() && effect.getPotionID() == Potion.fireResistance.id)
                || (!showWaterBreathing.get() && effect.getPotionID() == Potion.waterBreathing.id)
                || (!showInvisibility.get() && effect.getPotionID() == Potion.invisibility.id)
                || (!showBlindness.get() && effect.getPotionID() == Potion.blindness.id)
                || (!showNightVision.get() && effect.getPotionID() == Potion.nightVision.id)
                || (!showHunger.get() && effect.getPotionID() == Potion.hunger.id)
                || (!showWeakness.get() && effect.getPotionID() == Potion.weakness.id)
                || (!showPoison.get() && effect.getPotionID() == Potion.poison.id)
                || (!showWither.get() && effect.getPotionID() == Potion.wither.id)
                || (!showHealthBoost.get() && effect.getPotionID() == Potion.healthBoost.id)
                || (!showAbsorption.get() && effect.getPotionID() == Potion.absorption.id)
                || (!showSaturation.get() && effect.getPotionID() == Potion.saturation.id)

                || (!showPermanent.get() && effect.getIsPotionDurationMax()))).collect(Collectors.toList());
    }

    @Override
    protected float getHitBoxWidth() {
        return this.width;
    }

    @Override
    protected float getHitBoxHeight() {
        return this.height;
    }

    public enum AmplifierMode {
        ROMAN,
        ARABIC
    }

    public enum VerticalAlignmentMode {
        DOWN,
        UP
    }

    public enum SortingMode {
        ALPHABETICAL,
        DURATION,
        VANILLA
    }

    public static class PotionHUDTracker {

        public static final PotionHUDTracker INSTANCE = new PotionHUDTracker();

        public final Set<ElementPotionHUD> instances;

        private PotionHUDTracker() {
            this.instances = new HashSet<>();
        }

    }

}

File: src/main/kotlin/dev/isxander/evergreenhud/elements/impl/ElementPotionHUD.kt
/*
 * EvergreenHUD - A mod to improve your heads-up-display.
 * Copyright (c) isXander [2019 - 2022].
 *
 * This work is licensed under the GPL-3 License.
 * To view a copy of this license, visit https://www.gnu.org/licenses/gpl-3.0.en.html
 */

package dev.isxander.evergreenhud.elements.impl

import com.mojang.blaze3d.systems.RenderSystem
import dev.isxander.evergreenhud.elements.RenderOrigin
import dev.isxander.evergreenhud.elements.type.BackgroundElement
import dev.isxander.evergreenhud.elements.type.TextElement.*
import dev.isxander.evergreenhud.settings.color
import dev.isxander.evergreenhud.utils.*
import dev.isxander.evergreenhud.utils.elementmeta.ElementMeta
import dev.isxander.settxi.impl.*
import net.minecraft.client.resource.language.I18n
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.entity.effect.StatusEffectInstance
import net.minecraft.entity.effect.StatusEffects
import net.minecraft.util.Formatting
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry

@ElementMeta(id = "evergreenhud:potion_hud", name = "PotionHUD", category = "Player", description = "Display potions.")
class ElementPotionHUD : BackgroundElement() {
    var titleVisible by boolean(true) {
        name = "Visible"
        category = "Title"
        description = "If the title should be rendered."
    }

    var titleColor by color(Color.white) {
        name = "Color"
        category = "Title"
        description = "Color of the name of the potion effect."
        canHaveChroma = true
    }

    var titleStyle by option(TextStyle.SHADOW) {
        name = "Style"
        category = "Title"
        description = "In what style should the text be rendered."
    }

    var titleBold by boolean(true) {
        name = "Bold"
        category = "Title"
        description = "If the title should be bold."
    }

    var titleUnderlined by boolean(false) {
        name = "Underlined"
        category = "Title"
        description = "If the title should be underlined."
    }

    var titleItalic by boolean(false) {
        name = "Italic"
        category = "Title"
        description = "If the title should be slanted."
    }

    var amplifier by boolean(true) {
        name = "Show Amplifier"
        category = "Title"
        description = "Show the amplified of the potion effect."
    }

    var amplifierStyle by option(AmplifierText.ARABIC) {
        name = "Amplifier Style"
        category = "Title"
        description = "The style of the amplifier to use."
    }

    var showLvl1 by boolean(true) {
        name = "Show LVL 1"
        category = "Title"
        description = "Show the amplifier if it's level is 1."
    }


    var durationVisible by boolean(true) {
        name = "Visible"
        category = "Duration"
        description = "If the duration should be rendered."
    }

    var durationColor by color(Color.white) {
        name = "Color"
        category = "Duration"
        description = "Color of the name of the duration."
        canHaveChroma = true
    }

    var durationStyle by option(TextStyle.SHADOW) {
        name = "Style"
        category = "Duration"
        description = "In what style should the text be rendered."
    }

    var durationBold by boolean(true) {
        name = "Bold"
        category = "Duration"
        description = "If the duration should be bold."
    }

    var durationUnderlined by boolean(false) {
        name = "Underlined"
        category = "Duration"
        description = "If the duration should be underlined."
    }

    var durationItalic by boolean(false) {
        name = "Italic"
        category = "Duration"
        description = "If the duration should be slanted."
    }

    var permanentText by string("**:**") {
        name = "Permanent Text"
        category = "Duration"
        description = "The text to display when the effect is permanent."
    }

    var blinkingTime by int(5) {
        name = "Blinking Time"
        category = "Duration"
        description = "How many seconds have to remain before the duration starts to flash."
        range = 0..30
    }

    var blinkingSpeed by int(30) {
        name = "Blinking Speed"
        category = "Duration"
        description = "How fast the duration blinks."
        range = 10..45
    }

    var showIcon by boolean(true) {
        name = "Show Icon"
        category = "Miscellaneous"
        description = "Show the potion icon."
    }

    var sort by option(PotionSorting.DURATION) {
        name = "Sort"
        category = "Miscellaneous"
        description = "How the potion effects should be sorted."
    }

    var invertSort by boolean(false) {
        name = "Invert Sort"
        category = "Miscellaneous"
        description = "Inverts the sorting method."
    }

    var verticalSpacing by int(2) {
        name = "Vertical Spacing"
        category = "Miscellaneous"
        description = "How far apart each potion is."
        range = 1..9
    }

    val effectToggles = hashMapOf<Identifier, BooleanSetting>()

    init {
        for (potion in Registry.STATUS_EFFECT.toList()) {
            if (potion.isInstant) continue

            val name = I18n.translate(potion.translationKey)
            effectToggles[Registry.STATUS_EFFECT.getId(potion)!!] = boolean(true) {
                this.name = name
                category = "Toggles"
                description = "Display $name on the list when effected."
            }
        }
    }

    private val iconSize = 18

    private var w = hitboxWidth
    private var h = hitboxHeight
    override val hitboxWidth: Float
        get() = w
    override val hitboxHeight: Float
        get() = h

    override fun render(matrices: MatrixStack, renderOrigin: RenderOrigin) {
        RenderSystem.setShaderColor(1f, 1f, 1f, 1f)

        val x = position.rawX
        val y = position.rawY

        val potionEffects = arrayListOf<StatusEffectInstance>()
        if (mc.player == null) return
        potionEffects.addAll(
            mc.player!!.activeStatusEffects
                .filter { (effect, _) ->
                    effectToggles[Registry.STATUS_EFFECT.getId(effect)]?.get() ?: false
                }
                .map { it.value }
        )

        if (potionEffects.isEmpty()) {
            if (renderOrigin != RenderOrigin.GUI) return

            potionEffects.add(StatusEffectInstance(StatusEffects.STRENGTH, 50, 0))
            potionEffects.add(StatusEffectInstance(StatusEffects.ABSORPTION, 5, 0))
            potionEffects.add(StatusEffectInstance(StatusEffects.HASTE, 2000, 3))
        }

        when (sort) {
            PotionSorting.ALPHABETICAL -> potionEffects.sortBy { I18n.translate(it.translationKey) }
            PotionSorting.DURATION -> potionEffects.sortByDescending(StatusEffectInstance::getDuration)
        }

        if (invertSort) potionEffects.reverse()

        var yOff = 0f
        var xOff = 0f
        val yAmt = iconSize.toFloat() + verticalSpacing

        h = potionEffects.size * yAmt - verticalSpacing
        super.render(matrices, renderOrigin)
        w = 10f

        matrices.push()
        matrices.scale(position.scale, position.scale, 1f)
        for (effect in potionEffects) {
            var iconX = x
            iconX /= position.scale
            RenderSystem.setShaderColor(1f, 1f, 1f, 1f)
            if (showIcon) {
                val sprite = mc.statusEffectSpriteManager.getSprite(effect.effectType)
                RenderSystem.setShaderTexture(0, sprite.atlas.id)
                matrices.drawSprite(iconX, (y + yOff) / position.scale, 0f, 18f, 18f, sprite)

                xOff = iconSize * position.scale
                w = w.coerceAtLeast(xOff / position.scale)
            }
            if (titleVisible) {
                if (showIcon) xOff = (iconSize + 4) * position.scale
                val titleSb = StringBuilder()
                if (titleBold) titleSb.append(Formatting.BOLD)
                if (titleItalic) titleSb.append(Formatting.ITALIC)
                if (titleUnderlined) titleSb.append(Formatting.UNDERLINE)
                titleSb.append(I18n.translate(effect.translationKey))
                val amplifier = 1.coerceAtLeast(effect.amplifier + 1)
                if (this.amplifier && (amplifier != 1 || showLvl1)) {
                    titleSb.append(" ")
                    if (amplifierStyle === AmplifierText.ROMAN) titleSb.append(getRoman(amplifier))
                    else titleSb.append(amplifier)
                }
                val builtTitle = titleSb.toString()
                val titleWidth: Int = mc.textRenderer.getWidth(builtTitle)
                w = w.coerceAtLeast(xOff / position.scale + titleWidth)
                var titleX = x + xOff
                titleX /= position.scale
                var titleY = y + yOff
                if (!durationVisible) titleY += mc.textRenderer.fontHeight / 2f
                titleY /= position.scale
                drawString(
                    matrices,
                    builtTitle,
                    titleX,
                    titleY,
                    titleColor.rgba,
                    titleStyle == TextStyle.SHADOW,
                    false,
                    titleStyle == TextStyle.BORDER,
                    titleColor.chroma
                )
            }
            if (durationVisible) {
                if (showIcon) xOff = (iconSize + 4) * position.scale
                val durationSb = StringBuilder()
                if (durationBold) durationSb.append(Formatting.BOLD)
                if (durationItalic) durationSb.append(Formatting.ITALIC)
                if (durationUnderlined) durationSb.append(Formatting.UNDERLINE)
                if (effect.isPermanent) durationSb.append(permanentText) else durationSb.append(ticksToTime(effect.duration))
                val builtDuration = durationSb.toString()
                val durationWidth = mc.textRenderer.getWidth(builtDuration)
                w = w.coerceAtLeast(xOff / position.scale + durationWidth)
                var timeX = x + xOff
                timeX /= position.scale
                var timeY: Float = y + yOff + mc.textRenderer.fontHeight * position.scale + 1
                if (!titleVisible) timeY -= mc.textRenderer.fontHeight / 2f
                timeY /= position.scale
                if (effect.duration / 20f > blinkingTime || effect.duration % (50 - blinkingSpeed) <= (50 - blinkingSpeed) / 2f) {
                    drawString(
                        matrices,
                        builtDuration,
                        timeX,
                        timeY,
                        durationColor.rgba,
                        durationStyle == TextStyle.SHADOW,
                        false,
                        durationStyle == TextStyle.BORDER,
                        durationColor.chroma
                    )
                }
            }
            yOff += yAmt * position.scale
        }
        matrices.pop()
    }

    object AmplifierText : OptionContainer() {
        val ARABIC = option("Arabic", "Normal numbers.")
        val ROMAN = option("Roman", "Numbers like IX")
    }

    object PotionSorting : OptionContainer() {
        val ALPHABETICAL = option("Alphabetical", "Sort the potions by the alphabetical order of the name.")
        val DURATION = option("Duration", "Sort the potions by how long they have left.")
        val VANILLA = option("Vanilla", "Sort the potions in the way vanilla minecraft does.")
    }
}
